power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_se_solver.hpp

```diff
@@ -170,21 +167,277 @@ template <bool sym> class NewtonRaphsonSESolver {
+            data_gain_[data_idx_lu].qt_P_v() = data_gain_[data_idx_tranpose].q_Q_theta();
+            data_gain_[data_idx_lu].qt_Q_theta() = data_gain_[data_idx_tranpose].q_P_v();
+            data_gain_[data_idx_lu].qt_Q_v() = data_gain_[data_idx_tranpose].q_Q_v();
+        }
         // prefactorize
         sparse_solver_.prefactorize(data_gain_, perm_);
     }
 
+    NRSEJacobian power_flow_jacobian_i(RealTensor<sym> const& gs_minus_bc, RealTensor<sym> const& gc_plus_bs,
+                                       RealDiagonalTensor<sym> const& abs_u_self_inv,
+                                       RealValue<sym> const& calculated_p, RealValue<sym> const& calculated_q) {
+        auto jacobian = power_flow_jacobian_j(gs_minus_bc, gc_plus_bs, abs_u_self_inv);
+        jacobian.dP_dt -= RealTensor<sym>{calculated_q};
+        jacobian.dP_dv += dot(calculated_p, abs_u_self_inv);
+        jacobian.dQ_dt += RealTensor<sym>{calculated_p};
+        jacobian.dQ_dv += dot(calculated_q, abs_u_self_inv);
+        return jacobian;
+    }
+
+    NRSEJacobian power_flow_jacobian_j(RealTensor<sym> const& gs_minus_bc, RealTensor<sym> const& gc_plus_bs,
+                                       RealDiagonalTensor<sym> const& abs_u_other_inv) {
+        NRSEJacobian jacobian{};
+        jacobian.dP_dt = gs_minus_bc;
+        jacobian.dP_dv = -gc_plus_bs;
+        jacobian.dQ_dt = dot(gc_plus_bs, abs_u_other_inv);
+        jacobian.dQ_dv = dot(gs_minus_bc, abs_u_other_inv);
+        return jacobian;
+    }
+
+    void multiply_add_jacobian_blocks(NRSEGainBlock<sym>& block, NRSERhs<sym>& rhs_block, NRSEJacobian const& block_1,
+                                      NRSEJacobian const& block_2, PowerSensorCalcParam<sym> const& power_sensor,
+                                      RealValue<sym> const& calculated_power_p,
+                                      RealValue<sym> const& calculated_power_q) {
+        auto const w_p = diagonal_inverse(power_sensor.p_variance);
+        auto const w_q = diagonal_inverse(power_sensor.q_variance);
+        auto const del_power_p = calculated_power_p - real(power_sensor.value);
+        auto const del_power_q = calculated_power_q - imag(power_sensor.value);
+
+        // matrix multiplication of F_k^T . w_k . F_k
+        block.g_P_theta() += dot(w_p, block_1.dP_dt, block_2.dP_dt) + dot(w_q, block_1.dQ_dt, block_2.dQ_dt);
+        block.g_P_v() += dot(w_p, block_1.dP_dv, block_2.dP_dv) + dot(w_q, block_1.dQ_dv, block_2.dQ_dv);
+        block.g_Q_theta() += dot(w_p, block_1.dP_dt, block_2.dP_dv) + dot(w_q, block_1.dQ_dt, block_2.dQ_dv);
+        block.g_Q_v() += dot(w_p, block_1.dP_dv, block_2.dP_dt) + dot(w_q, block_1.dQ_dv, block_2.dQ_dt);
+
+        // matrix multiplication of F_k^T . w_k . (z - f(x))
+        rhs_block.eta_theta() += dot(w_p, block_1.dP_dt, del_power_p) + dot(w_q, block_1.dQ_dt, del_power_q);
+        rhs_block.eta_v() += dot(w_p, block_1.dP_dv, del_power_p) + dot(w_q, block_1.dQ_dv, del_power_q);
```

Comment:
while i feel like this could be implemented differently to prevent repeating yourself, this is very readable and directly translates to the overleaf document :+1: i understand what's happening and it looks clear